File: apps/slate-mac/Sources/SlateMac/SearchRecentsStore.swift
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

import Foundation

/// Per-vault persistence for the search overlay's recent-query list
/// (#876; HIG macos/searching.md:37 — "show recent searches … before
/// and during typing", :38 privacy — offer a way to clear them).
///
/// Structurally a `FileRecentsStore` sibling — LRU `add`, dedupe-on-load,
/// `maxEntries` cap, `maxFileBytes` bounded read, malformed-tolerant load,
/// WorkspaceStore temp-file-then-rename write discipline — with two
/// deliberate differences:
///
///  - **Entries are raw query strings** (`"budget 2026"`), not file paths
///    or command ids. Queries are vault-content-specific, so — like
///    `FileRecentsStore` and unlike the global `CommandPaletteRecentsStore`
///    — the store is **per vault**, at `<vault>/.slate/search-recents.json`
///    (beside `workspace.json` / `file-recents.json`, NOT the CLI-shared
///    `prefs.json`).
///  - **`clear()`** wipes the whole list in one call — the idle-state
///    "Clear" affordance the privacy note calls for. (`FileRecentsStore`
///    never needed a bulk clear; `CommandPaletteRecentsStore` exposes a
///    single-id `remove` instead.)
///
/// A malformed / oversized / vanished file always degrades to an empty
/// list — the overlay must open regardless of recents state (it falls
/// back to the "Type to search." idle hint).
///
/// Concurrency: load/save are synchronous and main-thread-friendly.
/// Single-writer (one user, one process), so `add` does
/// read → mutate → write each call with no locking.
struct SearchRecentsStore {
    /// Hard cap on persisted recent queries. Smaller than the file
    /// store's 50 (a compact overlay list, not a deep navigation
    /// history) but larger than a single screenful so a little scroll-
    /// back survives. Enforced on load (dedupe short-circuit) and on
    /// `add`.
    static let maxEntries = 20

    /// Upper bound on the recents file `load()` will read (mirrors the
    /// file/palette stores' #341 guard). 20 short query strings stay
    /// well under this; a file over it is corrupt or hostile and is
    /// treated as malformed (→ empty list).
    static let maxFileBytes: Int = 1 << 16  // 64 KiB

    let vaultRoot: URL

    /// `<vault>/.slate/search-recents.json` — same `.slate/` directory
    /// WorkspaceStore uses for `workspace.json`.
    var fileURL: URL {
        vaultRoot.appendingPathComponent(".slate/search-recents.json")
    }

    /// The on-disk list (most-recent-first), or empty on a missing /
    /// malformed / oversized / unreadable file. Dedupes on load and
    /// short-circuits at `maxEntries` — the same bounded-read + dedupe
    /// discipline as `FileRecentsStore.load()`; see
    /// `CommandPaletteRecentsStore.load()` for the full rationale on why
    /// one byte past the cap is read and why every failure arm falls to
    /// the empty list.
    func load() -> [String] {
        guard let handle = try? FileHandle(forReadingFrom: fileURL) else { return [] }
        defer { try? handle.close() }
        guard let data = try? handle.read(upToCount: Self.maxFileBytes + 1) else {
            return []
        }
        if data.count > Self.maxFileBytes {
            let message =
                "SearchRecentsStore: recents file exceeds the "
                + "\(Self.maxFileBytes)-byte threshold; treating as malformed."
            NSLog("%@", message)
            return []
        }
        guard let decoded = try? JSONDecoder().decode([String].self, from: data) else {
            return []
        }
        var seen = Set<String>()
        var deduped: [String] = []
        deduped.reserveCapacity(Self.maxEntries)
        for query in decoded {
            if seen.insert(query).inserted {
                deduped.append(query)
                if deduped.count >= Self.maxEntries { break }
            }
        }
        return deduped
    }

    /// Atomically overwrite the on-disk list. Temp-file-then-rename in
    /// `.slate/` (same volume ⇒ atomic) — the WorkspaceStore write
    /// discipline `FileRecentsStore` uses.
    func save(_ queries: [String]) throws {
        let dir = fileURL.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(queries)
        let tmp = dir.appendingPathComponent("search-recents.json.tmp-\(UUID().uuidString)")
        try data.write(to: tmp, options: [])
        _ = try FileManager.default.replaceItemAt(fileURL, withItemAt: tmp)
    }

    /// Add a query, moving an existing entry to the front (LRU).
    /// Returns the updated list, capped at `maxEntries`.
    @discardableResult
    func add(_ query: String) throws -> [String] {
        var entries = load()
        entries.removeAll { $0 == query }
        entries.insert(query, at: 0)
        if entries.count > Self.maxEntries {
            entries = Array(entries.prefix(Self.maxEntries))
        }
        try save(entries)
        return entries
    }

    /// Forget every remembered query (the idle-state "Clear" affordance,
    /// searching.md:38). Persists an empty list rather than deleting the
    /// file so a subsequent `load()` path stays identical.
    func clear() throws {
        try save([])
    }
}

File: apps/slate-windows/src/SlateWindows/Search/SearchRecentsStore.cs
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

using System.IO;
using System.Text.Json;

namespace SlateWindows.Search;

/// <summary>
/// Per-vault persistence for the search overlay's recent-query list —
/// the interoperating twin of mac's <c>SearchRecentsStore</c>
/// (<c>SearchRecentsStore.swift:33-124</c>), contract S14.
/// </summary>
/// <remarks>
/// <para>
/// The store is a <b>vault artifact</b> — <c>{vault}/.slate/search-recents.json</c>,
/// a JSON array of raw query strings, most-recent-first — and vaults
/// move between platforms, so this class must read what mac wrote and
/// write what mac reads, not merely imitate the shape.
/// </para>
/// <para>
/// Every read and write goes through the W1 anchored-vault discipline:
/// <see cref="AnchoredVaultStore"/> holds the vault root and its
/// <c>.slate</c> directory open, verifies both against external reparse
/// points, and renames the temp file by handle — the same fail-closed
/// mechanism <c>sidebar.json</c> and <c>workspace.json</c> use. The
/// file is never opened naively. No lock file is taken: like mac's
/// store, this is single-writer by design (one user, one process), and
/// <see cref="Add"/> is read → mutate → atomic-write per call.
/// </para>
/// <para>
/// A missing, malformed, oversized, or unreadable file degrades to an
/// <b>empty list, never a throw</b> — the overlay must open regardless
/// of recents state. Write failures are equally non-fatal, surfaced on
/// <see cref="LastSaveError"/> (the palette recents-store precedent);
/// callers re-read via <see cref="Load"/> so a failed
/// <see cref="Clear"/> keeps showing the still-persisted list rather
/// than pretending it was forgotten.
/// </para>
/// </remarks>
internal sealed class SearchRecentsStore
{
    /// <summary>Hard cap on persisted recent queries — mirror of mac's
    /// <c>maxEntries</c>. Enforced on load (dedupe, first occurrence
    /// wins, short-circuit at the cap) and on <see cref="Add"/>.</summary>
    internal const int MaxEntries = 20;

    /// <summary>Upper bound on the bytes <see cref="Load"/> will accept —
    /// mirror of mac's <c>maxFileBytes</c> (64 KiB). The bounded read
    /// detects the first byte past the cap from the open handle (no
    /// stat-then-read TOCTOU window) and treats a larger file as
    /// malformed.</summary>
    internal const int MaxFileBytes = 64 * 1024;

    private const string FileName = "search-recents.json";

    private static readonly JsonSerializerOptions WriteOptions = new()
    {
        WriteIndented = true,
    };

    private readonly string _vaultRoot;

    public SearchRecentsStore(string vaultRoot)
    {
        ArgumentException.ThrowIfNullOrEmpty(vaultRoot);
        _vaultRoot = Path.GetFullPath(vaultRoot);
    }

    /// <summary>The last persistence failure, or <see langword="null"/>.
    /// Non-fatal, but not invisible.</summary>
    public Exception? LastSaveError { get; private set; }

    /// <summary>
    /// The on-disk list (most-recent-first), or empty on a missing,
    /// malformed, oversized, or unreadable file. Dedupes on load and
    /// short-circuits at <see cref="MaxEntries"/>.
    /// </summary>
    /// <remarks>
    /// A JSON array containing a non-string element is treated as
    /// malformed in whole (empty list), matching mac's
    /// <c>JSONDecoder().decode([String].self)</c>, which rejects the
    /// entire file rather than skipping the entry.
    /// </remarks>
    public IReadOnlyList<string> Load()
    {
        try
        {
            using AnchoredVaultStore? store = AnchoredVaultStore.Open(
                _vaultRoot,
                createDirectory: false);
            byte[]? input = store?.ReadAllBytesBounded(FileName, MaxFileBytes);
            if (input is null)
            {
                return [];
            }

            string?[]? decoded = JsonSerializer.Deserialize<string?[]>(input);
            if (decoded is null)
            {
                return [];
            }

            var entries = new List<string>(decoded.Length);
            foreach (string? entry in decoded)
            {
                if (entry is null)
                {
                    return [];
                }

                entries.Add(entry);
            }

            return Sanitize(entries);
        }
        catch (Exception exception) when (
            exception is IOException or UnauthorizedAccessException or JsonException)
        {
            // FileSizeLimitExceededException (oversized) and
            // DirectoryNotFoundException (no vault root) are IOExceptions
            // and land here too.
            return [];
        }
    }

    /// <summary>
    /// Add a query, moving an existing entry to the front (LRU: remove
    /// equal, insert front, cap, atomic save). Returns the updated
    /// in-memory list, correct even when persistence failed.
    /// </summary>
    public IReadOnlyList<string> Add(string query)
    {
        ArgumentNullException.ThrowIfNull(query);
        var entries = Load().ToList();
        entries.RemoveAll(entry => string.Equals(entry, query, StringComparison.Ordinal));
        entries.Insert(0, query);
        if (entries.Count > MaxEntries)
        {
            entries.RemoveRange(MaxEntries, entries.Count - MaxEntries);
        }

        Save(entries);
        return entries;
    }

    /// <summary>
    /// Forget every remembered query. Persists an empty list rather
    /// than deleting the file, so the subsequent load path is identical
    /// (contract S14).
    /// </summary>
    public void Clear() => Save([]);

    private void Save(IReadOnlyList<string> queries)
    {
        try
        {
            using AnchoredVaultStore store = AnchoredVaultStore.Open(
                _vaultRoot,
                createDirectory: true)
                ?? throw new IOException("Could not anchor the search-recents directory.");
            // Pretty-printed like mac's encoder; the schema either side
            // reads is simply "a JSON array of strings".
            byte[] output = JsonSerializer.SerializeToUtf8Bytes(queries, WriteOptions);
            store.WriteAtomically(FileName, output);
            LastSaveError = null;
        }
        catch (Exception exception) when (
            exception is IOException or UnauthorizedAccessException)
        {
            LastSaveError = exception;
        }
    }

    private static IReadOnlyList<string> Sanitize(IReadOnlyList<string> queries)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var output = new List<string>(MaxEntries);
        foreach (string query in queries)
        {
            if (seen.Add(query))
            {
                output.Add(query);
                if (output.Count == MaxEntries)
                {
                    break;
                }
            }
        }

        return output;
    }
}
